core/src/main/java/com/scalar/db/transaction/consensuscommit/CoordinatorGroupCommitter.java

```diff
@@ -0,0 +1,125 @@
+package com.scalar.db.transaction.consensuscommit;
+
+import com.scalar.db.util.groupcommit.GroupCommitConfig;
+import com.scalar.db.util.groupcommit.GroupCommitter;
+import com.scalar.db.util.groupcommit.KeyManipulator;
+import java.util.Optional;
+import java.util.Random;
+
+public class CoordinatorGroupCommitter
+    extends GroupCommitter<String, String, String, String, Snapshot> {
+  CoordinatorGroupCommitter(GroupCommitConfig config) {
+    super("coordinator", config, new CoordinatorGroupCommitKeyManipulator());
+  }
+
+  public CoordinatorGroupCommitter(ConsensusCommitConfig config) {
+    this(
+        new GroupCommitConfig(
+            config.getCoordinatorGroupCommitSlotCapacity(),
+            config.getCoordinatorGroupCommitGroupSizeFixTimeoutMillis(),
+            config.getCoordinatorGroupCommitDelayedSlotMoveTimeoutMillis(),
+            config.getCoordinatorGroupCommitOldGroupAbortTimeoutMillis(),
+            config.getCoordinatorGroupCommitTimeoutCheckIntervalMillis(),
+            config.isCoordinatorGroupCommitMetricsMonitorLogEnabled()));
+  }
+
+  public static Optional<CoordinatorGroupCommitter> from(ConsensusCommitConfig config) {
+    if (config.isCoordinatorGroupCommitEnabled()) {
+      return Optional.of(new CoordinatorGroupCommitter(config));
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  static class CoordinatorGroupCommitKeyManipulator
+      implements KeyManipulator<String, String, String, String> {
+    private static final int PRIMARY_KEY_SIZE = 24;
+    private static final char DELIMITER = ':';
+    private static final int MAX_FULL_KEY_SIZE = 64;
+    private static final int MAX_CHILD_KEY_SIZE =
+        MAX_FULL_KEY_SIZE - PRIMARY_KEY_SIZE - 1 /* delimiter */;
+    private static final char[] CHARS_FOR_PRIMARY_KEY;
+    private static final int CHARS_FOR_PRIMARY_KEY_SIZE;
+
+    static {
+      int digitsLen = '9' - '0' + 1;
+      int upperCasesLen = 'Z' - 'A' + 1;
+      int lowerCasesLen = 'z' - 'a' + 1;
+      CHARS_FOR_PRIMARY_KEY = new char[digitsLen + upperCasesLen + lowerCasesLen];
+
+      int index = 0;
+      for (char c = '0'; c <= '9'; c++) {
+        CHARS_FOR_PRIMARY_KEY[index++] = c;
+      }
+      for (char c = 'A'; c <= 'Z'; c++) {
+        CHARS_FOR_PRIMARY_KEY[index++] = c;
+      }
+      for (char c = 'a'; c <= 'z'; c++) {
+        CHARS_FOR_PRIMARY_KEY[index++] = c;
+      }
+
+      CHARS_FOR_PRIMARY_KEY_SIZE = CHARS_FOR_PRIMARY_KEY.length;
+    }
+
+    // Use Random instead of ThreadLocalRandom in favor of global randomness.
+    private final Random random = new Random();
+
+    @Override
+    public String generateParentKey() {
```

Comment:
Random string generation might affect the performance, so it would be great if you could do a simple benchmark with multiple threads.

I previously encountered crazily slow random string generation when using an Apache Commons library, so I used a method taken from the benchbase one instead.
https://github.com/scalar-labs/scalardb-benchmarks/blob/master/src/main/java/com/scalar/db/benchmarks/ycsb/YcsbCommon.java#L109

So, it would be great if you could compare this with the one above.

Reply:
Thanks for sharing the concern. That makes sense.

I benchmarked the following combinations with [this commit](https://github.com/scalar-labs/scalardb/commit/3b2be71c72b6911655226143675fd92317c5bd2a).

a. Simple impl (the original `generateParentKey()` impl) + `Random` (as same as the current impl)
b. Faster impl (ported from `scalardb-benchmark`) + `Random`
c. Simple impl (the original `generateParentKey()` impl) + `ThreadLocalRandom`
d. Faster impl (ported from `scalardb-benchmark`) + `ThreadLocalRandom`

<img width="601" alt="image" src="https://github.com/scalar-labs/scalardb/assets/59043/6089ecfc-f79e-4121-97ed-e9342be0a066">

What I noticed from the result are:
- The implementation ported from `scalardb-benchmark` is about 10-15% faster than the original simple `generateParentKey()` impl when using `ThreadLocalRandom`
- Which to use `Random` or `ThreadLocalRandom` significantly affects the performance along with the number of threads. The average duration was 0.015 ms in case `a` with 16 threads which is the same number of the Linux machine's CPUs. The duration would increase when using a larger number of CPU machine

I think we should use `ThreadLocalRandom` just in case. When I benchmarked the group commit feature with `ThreadLocalRandom` before, I observed a lot of parent key conflicts at the beginning of benchmarking. I'll come up with a way to mitigate the issue and update this PR.

Reply:
@feeblefakie BTW, I checked the distribution of the random character values when using the implementation ported from `scalardb-benchmark`. It seems skewed compared to the original simple impl. I also observed a similar distribution when directly testing https://github.com/scalar-labs/scalardb-benchmarks/blob/master/src/main/java/com/scalar/db/benchmarks/ycsb/YcsbCommon.java#L109.

I think it's okay to generate just a random value, but the simple one would be better when generating unique identifiers.

- The implementation ported from `scalardb-benchmark`.
```
[000] 12170
[001] 21654
[002] 13412
[003] 23733
[004] 15577
[005] 26786
[006] 15973
[007] 56956
[008] 14013
[009] 26212
[010] 14705
[011] 107740
[012] 13730
[013] 28530
[014] 19633
[015] 106851
[016] 10670
[017] 29521
[018] 19581
[019] 61491
[020] 17591
[021] 46623
[022] 22265
[023] 48223
[024] 16174
[025] 58574
[026] 17782
[027] 44353
[028] 16499
[029] 69603
[030] 18314
[031] 40725
[032] 8898
[033] 73426
[034] 17933
[035] 47698
[036] 12817
[037] 81226
[038] 18393
[039] 30052
[040] 12386
[041] 64607
[042] 14837
[043] 53717
[044] 15901
[045] 73947
[046] 17029
[047] 30590
[048] 11894
[049] 34586
[050] 22686
[051] 52105
[052] 16454
[053] 43558
[054] 22372
[055] 33789
[056] 15510
[057] 24684
[058] 17335
[059] 28752
[060] 18036
[061] 29118
```

- The original simple implementation
```
[000] 39054
[001] 38329
[002] 38803
[003] 39094
[004] 38654
[005] 38724
[006] 38692
[007] 38661
[008] 38667
[009] 38810
[010] 38764
[011] 38945
[012] 38641
[013] 38578
[014] 38603
[015] 38749
[016] 38753
[017] 38709
[018] 38565
[019] 38706
[020] 38555
[021] 38993
[022] 38613
[023] 38477
[024] 38939
[025] 38521
[026] 38813
[027] 38752
[028] 38829
[029] 38728
[030] 38606
[031] 38984
[032] 38803
[033] 38611
[034] 39012
[035] 38577
[036] 38516
[037] 38485
[038] 38580
[039] 38982
[040] 38378
[041] 38650
[042] 38782
[043] 38494
[044] 38518
[045] 38574
[046] 38892
[047] 38815
[048] 38335
[049] 38681
[050] 38510
[051] 38572
[052] 38581
[053] 38473
[054] 38734
[055] 38822
[056] 39052
[057] 39031
[058] 38996
[059] 38902
[060] 38711
[061] 38620
```

Reply:
> When I benchmarked the group commit feature with ThreadLocalRandom before, I observed a lot of parent key conflicts at the beginning of benchmarking

This was my silly mistake. I stored the return value from `ThreadLocalRandom.current()`, but I should've not done that...

> Usages of this class should typically be of the form: ThreadLocalRandom.current().nextX(...) (where X is Int, Long, etc). When all usages are of this form, it is never possible to accidently share a ThreadLocalRandom across multiple threads.

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadLocalRandom.html

Fixed it in https://github.com/scalar-labs/scalardb/pull/1728/commits/755b91144b7ab8f165f53799a1963f723e0b2dac